### PR TITLE
Fix float to decimal convert overflow

### PR DESCRIPTION
**Pull request by coraxster**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category (leave one):
- Bug Fix

Short description (up to few sentences):
```sql
select toDecimal32(10000.1, 6); -- should be serverError 407
--2147.483648
select toDecimal64(10000.1, 18);  -- should be serverError 407
--9.223372036854775808
select toDecimal128(1000000000000000000000.1, 18);  -- should be serverError 407
0.000000000000000000
```

### CODE REVIEW

File: dbms/tests/queries/0_stateless/00700_decimal_casts.sql [hunk 1]
```diff
@@ -250,3 +250,7 @@ SELECT CAST(1/0, 'Decimal(38, 2)'); -- { serverError 407 }
 SELECT CAST(0/0, 'Decimal(9, 3)'); -- { serverError 407 }
 SELECT CAST(0/0, 'Decimal(18, 4)'); -- { serverError 407 }
 SELECT CAST(0/0, 'Decimal(38, 5)'); -- { serverError 407 }
+
+select toDecimal32(10000.1, 6); -- { serverError 407 }
+select toDecimal64(10000.1, 18); -- { serverError 407 }
+select toDecimal128(1000000000000000000000.1, 18); -- { serverError 407 }
```
Comment (4ertus2) [inline dbms/tests/queries/0_stateless/00700_decimal_casts.sql]: no empty line at the end

**Review by 4ertus2 — APPROVED** [review]

File: dbms/tests/queries/0_stateless/00700_decimal_casts.sql (hunk 1 above)
Comment (4ertus2) [inline dbms/tests/queries/0_stateless/00700_decimal_casts.sql]: Need empty line at the end of file

Comment (coraxster) [reply dbms/tests/queries/0_stateless/00700_decimal_casts.sql]: done, thank you!

File: dbms/src/DataTypes/DataTypesDecimal.h [hunk 2]
```diff
@@ -320,12 +320,26 @@ inline std::enable_if_t<IsDataTypeNumber<FromDataType> && IsDataTypeDecimal<ToDa
...
+            min_value = __int128(0x8000000000000000ll) << 64; // min __int128
+            max_value = (__int128(0x7fffffffffffffffll) << 64) + 0xffffffffffffffffll; // max __int128
+        }
+        else
+        {
+            min_value = std::numeric_limits<ToNativeType>::min();
+            max_value = std::numeric_limits<ToNativeType>::max();
+        }
+        ToNativeType converted_value = static_cast<ToNativeType>(value * ToDataType::getScaleMultiplier(scale));
+        if (converted_value == min_value || converted_value == max_value)
```
Comment (4ertus2) [inline dbms/src/DataTypes/DataTypesDecimal.h]: OK there's an error where we check isfinite(value) before multiplication and do not check the result. But why static_cast of INF returns min_value? Is it something documented in standard? If not we could just check the result.

```
auto out = value * ToDataType::getScaleMultiplier(scale);
if (!std::isfinite(value))
    throw Exception( ...
```

**Review by 4ertus2 — CHANGES_REQUESTED** [review]

File: dbms/src/DataTypes/DataTypesDecimal.h (hunk 2 above)
Comment (4ertus2) [reply dbms/src/DataTypes/DataTypesDecimal.h]: This code works
```
    if constexpr (std::is_floating_point_v<FromFieldType>)
    {
        if (!std::isfinite(value))
            throw Exception("Decimal convert overflow. Cannot convert infinity or NaN to decimal", ErrorCodes::DECIMAL_OVERFLOW);

        auto out = value * ToDataType::getScaleMultiplier(scale);
        if constexpr (std::is_same_v<ToNativeType, Int128>)
        {
            static constexpr __int128 min_int128 = __int128(0x8000000000000000ll) << 64;
            static constexpr __int128 max_int128 = (__int128(0x7fffffffffffffffll) << 64) + 0xffffffffffffffffll;
            if (out < min_int128 || out > max_int128)
                throw Exception("Decimal convert overflow. Float is out of Decimal range", ErrorCodes::DECIMAL_OVERFLOW);
        }
        else
        {
            if (out < std::numeric_limits<ToNativeType>::min() || out > std::numeric_limits<ToNativeType>::max())
                throw Exception("Decimal convert overflow. Float is out of Decimal range", ErrorCodes::DECIMAL_OVERFLOW);
        }
        return out;
    }
```

File: dbms/tests/queries/0_stateless/00700_decimal_casts.sql (hunk 1 above)
Comment (4ertus2) [inline dbms/tests/queries/0_stateless/00700_decimal_casts.sql]: We need 3 more lines with negatives.
```
select toDecimal32(-10000.1, 6); -- { serverError 407 }
￼select toDecimal64(-10000.1, 18); -- { serverError 407 }
 select toDecimal128(-1000000000000000000000.1, 18); -- { serverError 407 }
```

**Comment by coraxster** [comment]
Got some non-fixed boundary cases. Made a new PR.
https://github.com/yandex/ClickHouse/pull/5645

**Pull request merged**